matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/session/room/create/CreateLocalRoomStateEventsTask.kt

```diff
@@ -0,0 +1,298 @@
+                type = EventType.STATE_ROOM_POWER_LEVELS,
+                content = (createRoomBody.powerLevelContentOverride ?: PowerLevelsContent()).let {
+                    it.copy(
+                            ban = it.banOrDefault(),
+                            eventsDefault = it.eventsDefaultOrDefault(),
+                            invite = it.inviteOrDefault(),
+                            kick = it.kickOrDefault(),
+                            redact = it.redactOrDefault(),
+                            stateDefault = it.stateDefaultOrDefault(),
+                            usersDefault = it.usersDefaultOrDefault(),
+                    )
+                }.toContent(),
+        )
+        add(powerLevelsContent)
+    }
+
+    /**
+     * Generate the local room member state events related to the given user ids, if any.
+     */
+    private suspend fun MutableList<Event>.createRoomMemberEvents(userIds: List<String>) = apply {
+        val memberEvents = userIds
+                .mapNotNull { tryOrNull { userService.resolveUser(it) } }
+                .map { user ->
+                    createLocalStateEvent(
+                            type = EventType.STATE_ROOM_MEMBER,
+                            content = RoomMemberContent(
+                                    isDirect = createRoomBody.isDirect.takeUnless { user.userId == roomCreatorUserId }.orFalse(),
+                                    membership = if (user.userId == roomCreatorUserId) Membership.JOIN else Membership.INVITE,
+                                    displayName = user.displayName,
+                                    avatarUrl = user.avatarUrl
+                            ).toContent(),
+                            stateKey = user.userId
+                    )
+                }
+        addAll(memberEvents)
+    }
+
+    /**
+     * Generate the local state events related to the given third party invites, if any.
+     */
+    private fun MutableList<Event>.createRoomThreePidEvents() = apply {
+        val threePidEvents = createRoomBody.invite3pids.orEmpty().map { body ->
+            val localThirdPartyInviteEvent = createLocalStateEvent(
+                    type = EventType.LOCAL_STATE_ROOM_THIRD_PARTY_INVITE,
+                    content = LocalRoomThirdPartyInviteContent(
+                            isDirect = createRoomBody.isDirect.orFalse(),
+                            membership = Membership.INVITE,
+                            displayName = body.address,
+                            thirdPartyInvite = body.toThreePid()
+                    ).toContent(),
+            )
+            val thirdPartyInviteEvent = createLocalStateEvent(
+                    type = EventType.STATE_ROOM_THIRD_PARTY_INVITE,
+                    content = RoomThirdPartyInviteContent(body.address, null, null, null).toContent(),
+            )
+            listOf(localThirdPartyInviteEvent, thirdPartyInviteEvent)
+        }.flatten()
+        addAll(threePidEvents)
+    }
```

Comment:
Maybe rewrite to:

```kotlin
    private fun MutableList<Event>.createRoomThreePidEvents() {
        createRoomBody.invite3pids.orEmpty().forEach { body ->
            val localThirdPartyInviteEvent = createLocalStateEvent(
                    type = EventType.LOCAL_STATE_ROOM_THIRD_PARTY_INVITE,
                    content = LocalRoomThirdPartyInviteContent(
                            isDirect = createRoomBody.isDirect.orFalse(),
                            membership = Membership.INVITE,
                            displayName = body.address,
                            thirdPartyInvite = body.toThreePid()
                    ).toContent(),
            )
            val thirdPartyInviteEvent = createLocalStateEvent(
                    type = EventType.STATE_ROOM_THIRD_PARTY_INVITE,
                    content = RoomThirdPartyInviteContent(
                            displayName = body.address,
                            keyValidityUrl = null,
                            publicKey = null,
                            publicKeys = null
                    ).toContent(),
            )
            add(localThirdPartyInviteEvent)
            add(thirdPartyInviteEvent)
        }
    }
```

for clarity

Reply:
💯 I don't know why I made it complicated! https://github.com/vector-im/element-android/pull/6127/commits/cac4df7d664b1e081f9c9bb36d217afbb410b5c9